### PR TITLE
Support variable instructions for locals

### PR DESCRIPTION
**Pull request by ixahedron**
Problem: we currently have extremely preliminary support for params and locals, but we don't support variable instructions at all.

Solution: introduced another `StateT` layer to our `EngineM` stack to house local values.
Supported instructions that manipulate locals, namely `local.get`, `local.set`, `local.tee`
NB: binary encoding is only supported for local instructions that refer to locals by indices, not by names

Also supported `drop`.

### CODE REVIEW

File: Wasm/Engine.lean [hunk 1]
```diff
@@ -346,14 +363,33 @@ mutual
       | .ofNat   _ => x <<< ((bs : Int) - k)
       | .negSucc _ => x >>> ((bs : Int) + k)
     | .rotr _ _ _ => throwEE .typecheck_failed
+    | .local_get (.by_index idx) => do match (←getLocals).get? idx with
+      | .some ⟨_, .some n, _⟩ => push $ .num n
+      | _ => throwEE $ .local_with_given_id_missing idx
+    | .local_get (.by_name name) => do
+      -- TODO: names are erased in production. See what do we want to do with this code path.
```
Comment (cognivore) [inline Wasm/Engine.lean:370]: See previous comment.

**Review by cognivore — APPROVED** [review]
LGTM + see notes, as for practical correctness, binary tests will tell.

**Comment by ixahedron** [comment]
@cognivore I added support for encoding `local.get by_name` and also refactored some `Bytes.lean` functions to do that properly. This might interfere with your own work, sorry about that! But I believe it's worthwhile work. If you want, we can collab on moving some of the work to your branch instead of leaving it here if it's easier to rebase that way.

**Comment by cognivore** [comment]
I'll review new.

File: Wasm/Bytes.lean [hunk 2]
```diff
@@ -36,36 +35,30 @@ def ttoi (x : Type') : UInt8 :=
...
-    let header := b 0x60 ++ uLeb128 params.length
-    let res := params.foldl Append.append header
-    res ++ (match x.results with -- TODO: test multi-result functions
-    | List.nil => b 0x00
-    | ts => b 0x7b ++ uLeb128 ts.length ++ flatten (ts.map $ b ∘ ttoi)
-    )
-  sigs.foldl
-    Append.append $
-      b 0x01 ++ uLeb128 (1 + totalLength sigs) ++ uLeb128 sigs.length
+def mkVec (xs : List α) (xtobs : α → ByteArray) : ByteArray :=
```
Comment (cognivore) [inline Wasm/Bytes.lean:41]: Sometimes we need to flatten with indecies.

```
def flattenWithIndecies : List ByteArray → ByteArray
  | [] => b0
  | bs => (bs.foldl (fun (acc, n) x => match x.data with
    | #[] => (acc, n)
    | _x => ((uLeb128 n ++ x ++ acc), n + 1)) (b0, 0)).1

-- Same as extractModIdentifier, but maps a list of functions into a length-prefixed wasm array.
def extractFuncIdentifiers (fs : List Func) : ByteArray :=
  let fbs :=
    flattenWithIndecies $ fs.map extractFuncIdentifier
  if fbs.size = 0 then
    b0
  else
    funcHeader ++ (lindex $ uLeb128 fs.length ++ fbs)
```

`mkVec` helps a lot, but I think that it's not generic. 

If only tthere was a generalisation for such folds. Let's read on and find out.

**Review by cognivore — COMMENTED** [review]
Yay! I'm very happy about locals now! I'll adjust my stuff next week, don't worry about it.

Of course, I assume that tests are green.

**Comment by ixahedron** [comment]
They are, if you mean the hopefully soon conclusively removed mess in Main.lean 😁

**Pull request merged**